File: agent/gate.py
"""Deterministic policy gate. Not an LLM. Sits before the shared channel."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from agent.orgs import identity_markers, raw_id_markers

ALLOWED_FIELDS = frozenset(
    {
        "org_id",
        "batch_id",
        "product_type",
        "aggregate_count",
        "risk_level",
        "supplier_id",
        "component_id",
        "allowed_alternative",
        "blocked",
        "reason",
        "id",
        "claim",
        "status",
        "from_org",
        "motion_id",
        "seconded",
    }
)

PROHIBITED_FIELDS = frozenset(
    {
        "customer_name",
        "customer_address",
        "customer_id",
        "email",
        "phone",
        "name",
        "address",
        "names",
        "addresses",
        "customer_ids",
        "customers",
    }
)

IDENTITY_MARKERS = identity_markers()
RAW_ID_MARKERS = raw_id_markers()

SHAREABLE_ID_PREFIXES = frozenset({"bat", "sup", "cell"})
_NAME_WORD = re.compile(
    r"[A-ZÄÖÜÁÉÍÓÚÀÈÌÒÙÑ][a-zäöüßáéíóúàèìòùñç]+"
    r"(?:['’-][A-Za-zÄÖÜäöüßÁÉÍÓÚáéíóúÑñç]+)*"
)
_ID_TOKEN = re.compile(r"\b([A-Za-z]{1,4})-(\d{3,})\b")
NOT_NAME_WORDS = frozenset(
    {
        "the",
        "and",
        "for",
        "from",
        "with",
        "this",
        "that",
        "returned",
        "return",
        "returns",
        "customer",
        "customers",
        "agent",
        "file",
        "batch",
        "shared",
        "write",
        "wireless",
        "earbuds",
        "battery",
        "swollen",
        "retailer",
        "supplier",
        "product",
        "claim",
        "status",
        "identity",
        "disclosure",
        "private",
        "record",
        "high",
        "medium",
        "low",
        "risk",
        "screen",
        "cracked",
        "items",
    }
)


@dataclass(frozen=True)
class GateDecision:
    allowed: bool
    reason: str | None
    allowed_alternative: str | None
    payload: dict[str, Any]

    def to_dict(self) -> dict[str, Any]:
        if self.allowed:
            return {"blocked": False, **self.payload}
        return {
            "blocked": True,
            "reason": self.reason,
            "allowed_alternative": self.allowed_alternative,
        }


def _walk_keys(value: Any) -> list[str]:
    keys: list[str] = []
    if isinstance(value, dict):
        for key, nested in value.items():
            keys.append(str(key).lower())
            keys.extend(_walk_keys(nested))
    elif isinstance(value, list):
        for item in value:
            keys.extend(_walk_keys(item))
    return keys


def _as_text(value: Any) -> str:
    if isinstance(value, str):
        return value.lower()
    if isinstance(value, (int, float, bool)):
        return ""
    return str(value).lower()


def _drop_prohibited(value: Any) -> Any:
    if isinstance(value, dict):
        return {
            key: _drop_prohibited(nested)
            for key, nested in value.items()
            if str(key).lower() not in PROHIBITED_FIELDS
        }
    if isinstance(value, list):
        return [_drop_prohibited(item) for item in value]
    return value


def _has_shareable(payload: dict[str, Any]) -> bool:
    if payload.get("batch_id") or payload.get("product_type") or payload.get("risk_level"):
        return True
    claim = payload.get("claim")
    if isinstance(claim, str) and claim.strip():
        return True
    if payload.get("aggregate_count") is not None:
        return True
    return False


def _identity_reason(prohibited: list[str]) -> str:
    if any(key in {"customer_id", "customer_ids"} for key in prohibited):
        return "RAW_IDENTIFIERS"
    if any(key in {"customer_name", "name", "names", "customers"} for key in prohibited):
        return "IDENTITY_DISCLOSURE"
    return "PROHIBITED_FIELDS"


def _walk_strings(value: Any) -> list[str]:
    if isinstance(value, str):
        return [value]
    if isinstance(value, dict):
        found: list[str] = []
        for nested in value.values():
            found.extend(_walk_strings(nested))
        return found
    if isinstance(value, list):
        found = []
        for item in value:
            found.extend(_walk_strings(item))
        return found
    return []


def _name_hits(text: str) -> list[str]:
    hits: list[str] = []
    matches = list(_NAME_WORD.finditer(text))
    index = 0
    while index < len(matches):
        run = [matches[index]]
        cursor = index + 1
        while cursor < len(matches):
            gap = text[run[-1].end() : matches[cursor].start()]
            if gap.strip() != "":
                break
            run.append(matches[cursor])
            cursor += 1
        if len(run) >= 2:
            tokens = [match.group(0) for match in run]
            if not any(token.lower() in NOT_NAME_WORDS for token in tokens):
                hits.append(" ".join(tokens))
        index += len(run)
    return list(dict.fromkeys(hits))


def _id_hits(text: str) -> list[str]:
    hits: list[str] = []
    for match in _ID_TOKEN.finditer(text):
        if match.group(1).lower() in SHAREABLE_ID_PREFIXES:
            continue
        hits.append(match.group(0))
    return list(dict.fromkeys(hits))


def find_identity(value: Any) -> list[str]:
    found: list[str] = []
    blob = _as_text(value)
    for marker in IDENTITY_MARKERS:
        if marker in blob:
            found.append(marker)
    for text in _walk_strings(value):
        for hit in _name_hits(text):
            found.append(hit.lower())
    return list(dict.fromkeys(found))


def find_identifiers(value: Any) -> list[str]:
    found: list[str] = []
    blob = _as_text(value)
    for marker in RAW_ID_MARKERS:
        if marker in blob:
            found.append(marker)
    for text in _walk_strings(value):
        for hit in _id_hits(text):
            found.append(hit.lower())
    return list(dict.fromkeys(found))


def redact_claim(claim: str) -> str:
    """Replace names and customer ids with a generic customer. Keep the rest."""
    slot = "a customer"
    text = claim
    markers = list(
        dict.fromkeys(
            [
                *find_identifiers(claim),
                *find_identity(claim),
                *_id_hits(claim),
                *_name_hits(claim),
            ]
        )
    )
    for marker in sorted(markers, key=len, reverse=True):
        text = re.sub(
            rf"\bcustomer\s+{re.escape(marker)}\b",
            slot,
            text,
            flags=re.IGNORECASE,
        )
        text = re.sub(re.escape(marker), slot, text, flags=re.IGNORECASE)
    text = re.sub(
        rf"(?:{re.escape(slot)})(?:\s+and\s+{re.escape(slot)})+",
        "customers",
        text,
        flags=re.IGNORECASE,
    )
    text = re.sub(
        rf"(?:{re.escape(slot)})(?:\s+{re.escape(slot)})+",
        slot,
        text,
        flags=re.IGNORECASE,
    )
    text = re.sub(r"\s+", " ", text)
    text = re.sub(r"\s+([.,;:])", r"\1", text)
    text = text.strip(" ,;:")
    if not text:
        return ""
    return text[0].upper() + text[1:]


def inspect(payload: dict[str, Any]) -> GateDecision:
    """Allow only policy-shareable fields onto the shared channel."""
    if not isinstance(payload, dict):
        return GateDecision(
            allowed=False,
            reason="INVALID_PAYLOAD",
            allowed_alternative="aggregate_count",
            payload={},
        )

    keys = _walk_keys(payload)
    prohibited = [key for key in keys if key in PROHIBITED_FIELDS]
    cleaned = _drop_prohibited(payload)
    if not isinstance(cleaned, dict):
        cleaned = {}

    if find_identity(cleaned):
        return GateDecision(
            allowed=False,
            reason="IDENTITY_DISCLOSURE",
            allowed_alternative="batch_return",
            payload={},
        )
    if find_identifiers(cleaned):
        return GateDecision(
            allowed=False,
            reason="RAW_IDENTIFIERS",
            allowed_alternative="batch_return",
            payload={},
        )

    extra = [key for key in _walk_keys(cleaned) if key not in ALLOWED_FIELDS]
    if extra:
        return GateDecision(
            allowed=False,
            reason="PROHIBITED_FIELDS",
            allowed_alternative="aggregate_count",
            payload={},
        )

    if prohibited:
        if not _has_shareable(cleaned):
            return GateDecision(
                allowed=False,
                reason=_identity_reason(prohibited),
                allowed_alternative="batch_return",
                payload={},
            )
        return GateDecision(
            allowed=True,
            reason=None,
            allowed_alternative=None,
            payload=cleaned,
        )

    return GateDecision(allowed=True, reason=None, allowed_alternative=None, payload=payload)


def scan(payload: dict[str, Any]) -> dict[str, Any]:
    """Same decision as inspect, plus the fields the demo can draw."""
    decision = inspect(payload)
    keys = _walk_keys(payload) if isinstance(payload, dict) else []
    prohibited = sorted({key for key in keys if key in PROHIBITED_FIELDS})
    extra = sorted(
        {
            key
            for key in keys
            if key not in ALLOWED_FIELDS and key not in PROHIBITED_FIELDS
        }
    )
    identity_hits = find_identity(payload) if isinstance(payload, dict) else []
    identifier_hits = find_identifiers(payload) if isinstance(payload, dict) else []
    kept = sorted(decision.payload.keys()) if decision.allowed else []
    return {
        "allowed": decision.allowed,
        "reason": decision.reason,
        "allowed_alternative": decision.allowed_alternative,
        "payload": decision.payload if decision.allowed else {},
        "keys": keys,
        "prohibited_fields": prohibited,
        "extra_fields": extra,
        "identity_hits": identity_hits,
        "identifier_hits": identifier_hits,
        "dropped_fields": sorted(set(prohibited) | set(extra)),
        "kept_fields": kept,
    }

File: web/src/lib/gate.ts
/** Same policy as agent/gate.py. Runs in Next so Vercel does not need uv. */

const ALLOWED_FIELDS = new Set([
  "org_id",
  "batch_id",
  "product_type",
  "aggregate_count",
  "risk_level",
  "supplier_id",
  "component_id",
  "allowed_alternative",
  "blocked",
  "reason",
  "id",
  "claim",
  "status",
  "from_org",
  "motion_id",
  "seconded",
]);

const PROHIBITED_FIELDS = new Set([
  "customer_name",
  "customer_address",
  "customer_id",
  "email",
  "phone",
  "name",
  "address",
  "names",
  "addresses",
  "customer_ids",
  "customers",
]);

const IDENTITY_MARKERS = [
  "anna müller",
  "anna mueller",
  "anna muller",
  "müller",
  "mueller",
  "peter schmidt",
  "schmidt",
  "leila hassan",
  "hassan",
  "jonas weber",
  "weber",
  "rollbergstraße 12, berlin",
  "rollbergstrasse 12, berlin",
  "sonnenallee 8, berlin",
  "kottbusser damm 4, berlin",
  "karl-marx-straße 1, berlin",
  "karl-marx-strasse 1, berlin",
];

const RAW_ID_MARKERS = ["c-1001", "c-1002", "c-1003", "c-1099"];
const SHAREABLE_ID_PREFIXES = new Set(["bat", "sup", "cell"]);
const NAME_WORD =
  /[A-ZÄÖÜÁÉÍÓÚÀÈÌÒÙÑ][a-zäöüßáéíóúàèìòùñç]+(?:['’-][A-Za-zÄÖÜäöüßÁÉÍÓÚáéíóúÑñç]+)*/g;
const ID_TOKEN = /\b([A-Za-z]{1,4})-(\d{3,})\b/g;
const NOT_NAME_WORDS = new Set([
  "the",
  "and",
  "for",
  "from",
  "with",
  "this",
  "that",
  "returned",
  "return",
  "returns",
  "customer",
  "customers",
  "agent",
  "file",
  "batch",
  "shared",
  "write",
  "wireless",
  "earbuds",
  "battery",
  "swollen",
  "retailer",
  "supplier",
  "product",
  "claim",
  "status",
  "identity",
  "disclosure",
  "private",
  "record",
  "high",
  "medium",
  "low",
  "risk",
  "screen",
  "cracked",
  "items",
]);

export type GateDecision = {
  allowed: boolean;
  reason: string | null;
  allowed_alternative: string | null;
  payload: Record<string, unknown>;
};

function walkKeys(value: unknown): string[] {
  if (Array.isArray(value)) return value.flatMap(walkKeys);
  if (value && typeof value === "object") {
    return Object.entries(value as Record<string, unknown>).flatMap(([key, nested]) => [
      key.toLowerCase(),
      ...walkKeys(nested),
    ]);
  }
  return [];
}

function asText(value: unknown): string {
  if (typeof value === "string") return value.toLowerCase();
  if (typeof value === "number" || typeof value === "boolean") return "";
  return JSON.stringify(value).toLowerCase();
}

function walkStrings(value: unknown): string[] {
  if (typeof value === "string") return [value];
  if (Array.isArray(value)) return value.flatMap(walkStrings);
  if (value && typeof value === "object") {
    return Object.values(value as Record<string, unknown>).flatMap(walkStrings);
  }
  return [];
}

function nameHits(text: string): string[] {
  const matches = [...text.matchAll(NAME_WORD)];
  const hits: string[] = [];
  let index = 0;
  while (index < matches.length) {
    const run = [matches[index]];
    let cursor = index + 1;
    while (cursor < matches.length) {
      const prev = run[run.length - 1];
      const next = matches[cursor];
      const gap = text.slice(prev.index! + prev[0].length, next.index);
      if (gap.trim() !== "") break;
      run.push(next);
      cursor += 1;
    }
    if (run.length >= 2) {
      const tokens = run.map((match) => match[0]);
      if (!tokens.some((token) => NOT_NAME_WORDS.has(token.toLowerCase()))) {
        hits.push(tokens.join(" "));
      }
    }
    index += run.length;
  }
  return [...new Set(hits)];
}

function idHits(text: string): string[] {
  const hits: string[] = [];
  for (const match of text.matchAll(ID_TOKEN)) {
    if (SHAREABLE_ID_PREFIXES.has(match[1].toLowerCase())) continue;
    hits.push(match[0]);
  }
  return [...new Set(hits)];
}

export function findIdentity(value: unknown): string[] {
  const found: string[] = [];
  const blob = asText(value);
  for (const marker of IDENTITY_MARKERS) {
    if (blob.includes(marker)) found.push(marker);
  }
  for (const text of walkStrings(value)) {
    for (const hit of nameHits(text)) found.push(hit.toLowerCase());
  }
  return [...new Set(found)];
}

export function findIdentifiers(value: unknown): string[] {
  const found: string[] = [];
  const blob = asText(value);
  for (const marker of RAW_ID_MARKERS) {
    if (blob.includes(marker)) found.push(marker);
  }
  for (const text of walkStrings(value)) {
    for (const hit of idHits(text)) found.push(hit.toLowerCase());
  }
  return [...new Set(found)];
}

function dropProhibited(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(dropProhibited);
  if (value && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const [key, nested] of Object.entries(value as Record<string, unknown>)) {
      if (PROHIBITED_FIELDS.has(key.toLowerCase())) continue;
      out[key] = dropProhibited(nested);
    }
    return out;
  }
  return value;
}

function hasShareable(payload: Record<string, unknown>): boolean {
  if (payload.batch_id || payload.product_type || payload.risk_level) return true;
  if (typeof payload.claim === "string" && payload.claim.trim()) return true;
  return payload.aggregate_count != null;
}

function identityReason(prohibited: string[]): string {
  if (prohibited.some((key) => key === "customer_id" || key === "customer_ids")) {
    return "RAW_IDENTIFIERS";
  }
  if (
    prohibited.some((key) =>
      ["customer_name", "name", "names", "customers"].includes(key),
    )
  ) {
    return "IDENTITY_DISCLOSURE";
  }
  return "PROHIBITED_FIELDS";
}

export function redactClaim(claim: string): string {
  const slot = "a customer";
  let text = claim;
  const markers = [
    ...new Set([
      ...findIdentifiers(claim),
      ...findIdentity(claim),
      ...idHits(claim),
      ...nameHits(claim),
    ]),
  ];
  for (const marker of markers.sort((a, b) => b.length - a.length)) {
    const escaped = marker.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    text = text.replace(new RegExp(`\\bcustomer\\s+${escaped}\\b`, "ig"), slot);
    text = text.replace(new RegExp(escaped, "ig"), slot);
  }
  text = text.replace(new RegExp(`(?:${slot})(?:\\s+and\\s+${slot})+`, "ig"), "customers");
  text = text.replace(new RegExp(`(?:${slot})(?:\\s+${slot})+`, "ig"), slot);
  text = text.replace(/\s+/g, " ").replace(/\s+([.,;:])/g, "$1").replace(/^[ ,;:]+|[ ,;:]+$/g, "");
  if (!text) return "";
  return text[0].toUpperCase() + text.slice(1);
}

export function inspect(payload: Record<string, unknown>): GateDecision {
  const keys = walkKeys(payload);
  const prohibited = keys.filter((key) => PROHIBITED_FIELDS.has(key));
  const cleanedRaw = dropProhibited(payload);
  const cleaned =
    cleanedRaw && typeof cleanedRaw === "object" && !Array.isArray(cleanedRaw)
      ? (cleanedRaw as Record<string, unknown>)
      : {};

  if (findIdentity(cleaned).length) {
    return {
      allowed: false,
      reason: "IDENTITY_DISCLOSURE",
      allowed_alternative: "batch_return",
      payload: {},
    };
  }
  if (findIdentifiers(cleaned).length) {
    return {
      allowed: false,
      reason: "RAW_IDENTIFIERS",
      allowed_alternative: "batch_return",
      payload: {},
    };
  }
  const extra = walkKeys(cleaned).filter((key) => !ALLOWED_FIELDS.has(key));
  if (extra.length) {
    return {
      allowed: false,
      reason: "PROHIBITED_FIELDS",
      allowed_alternative: "aggregate_count",
      payload: {},
    };
  }
  if (prohibited.length && !hasShareable(cleaned)) {
    return {
      allowed: false,
      reason: identityReason(prohibited),
      allowed_alternative: "batch_return",
      payload: {},
    };
  }
  if (prohibited.length) {
    return {
      allowed: true,
      reason: null,
      allowed_alternative: null,
      payload: cleaned,
    };
  }
  return {
    allowed: true,
    reason: null,
    allowed_alternative: null,
    payload,
  };
}

export function scan(payload: Record<string, unknown>) {
  const decision = inspect(payload);
  const keys = walkKeys(payload);
  const prohibited = [...new Set(keys.filter((key) => PROHIBITED_FIELDS.has(key)))].sort();
  const extra = [
    ...new Set(
      keys.filter((key) => !ALLOWED_FIELDS.has(key) && !PROHIBITED_FIELDS.has(key)),
    ),
  ].sort();
  return {
    allowed: decision.allowed,
    reason: decision.reason,
    allowed_alternative: decision.allowed_alternative,
    payload: decision.allowed ? decision.payload : {},
    keys,
    prohibited_fields: prohibited,
    extra_fields: extra,
    identity_hits: findIdentity(payload),
    identifier_hits: findIdentifiers(payload),
    dropped_fields: [...new Set([...prohibited, ...extra])].sort(),
    kept_fields: decision.allowed ? Object.keys(decision.payload).sort() : [],
  };
}

export const SHAREABLE_RETURN: Record<string, unknown> = {
  id: "m2",
  from_org: "org_a",
  org_id: "org_a",
  batch_id: "BAT-042",
  product_type: "wireless earbuds",
  risk_level: "high",
  claim: "A customer returned the earbuds. Battery swollen.",
};
